construct shapefiles in memory

donogst opened the pull request:
Fixes #885 
dotspatial.data.tests/zipexporttests.cs shows example of how to create the archive as well as testing that the contents of the archive match those of a regular save operation.
I think it has squashed correctly if for some reason it has not worked, let me know and I will try again

### Checklist

- [ y] I have included examples or tests
- [ y] I have updated the change log
- [ y] I am listed in the CONTRIBUTORS file
- [ y] I have cleaned up the commit history (use rebase and squash)

### Changes proposed in this pull request:

- construct shapefiles in memory only packaged in zip archive for download
- refactor of shapefile classes 
- shape export tests

mogikanin commented:
We can't use 3-rd party dependencies in DotSpatial core libraries. 
But we can split your changes into several parts:
1. Refactorings for support writing into streams.
2. Separate plugin\assembly with `ExportZipFile(string ShapefileName)` functionality. We can include it into DS repo as plugin, and later publish it as separate assembly at nuget.

donogst commented:
Hi, ok, I had seen that library elsewhere but obviously not in core.
Another option would be to upgrade to framework 4.5 which has native
support but that may be a large change for a small feature unless it is
already planned to use 4.5
On 20 Nov 2016 08:49, "Max Miroshnikov" <notifications@github.com> wrote:

> We can't use 3-rd party dependencies in DotSpatial core libraries.
> But we can split your changes into several parts:
> 1. Refactorings for support writing into streams.
> 2. Separate plugin\assembly with ExportZipFile(string ShapefileName)
> functionality. We can include it into DS repo as plugin, and later publish
> it as separate assembly at nuget.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DotSpatial/DotSpatial/pull/890#issuecomment-261766353>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFFvJ1OgKMWMs6_MPpkj_RZceaKqR1Zbks5rAAmggaJpZM4K3aeu>
> .
>

mogikanin commented:
@donogst , would you like to refactor your changes, because in current state we can't merge PR?

donogst commented:
Hi Max,

I have submitted a new pull request, rebased and squashed. Hope its all good

On 15 December 2016 at 14:36, Stephen Donoghue <donogst@googlemail.com>
wrote:

> I will try and do this by Monday.  The refactor should be easy enough
> On 15 Dec 2016 14:14, "Max Miroshnikov" <notifications@github.com> wrote:
>
>> @donogst <https://github.com/donogst> , would you like to refactor your
>> changes, because in current state we can't merge PR?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/DotSpatial/DotSpatial/pull/890#issuecomment-267337017>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AFFvJy5ujadtyk62h6jiQaLwDr-R79DBks5rIUsxgaJpZM4K3aeu>
>> .
>>
>

mogikanin commented:
@donogst I think you forgot to submit it as pull request. I see it in your branches, but not in pull requests.

donogst commented:
I had sent the request from my branch to my master. I have updated my
master which I think has updated my original PR
On 8 Jan 2017 09:09, "Max Miroshnikov" <notifications@github.com> wrote:

> @donogst <https://github.com/donogst> I think you forgot to submit it as
> pull request. I see it in your branches, but not in pull requests.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DotSpatial/DotSpatial/pull/890#issuecomment-271139406>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFFvJ_QF3yb-CiopVtZEgFLBFOc9M8a5ks5rQKe3gaJpZM4K3aeu>
> .
>

mogikanin commented:
Steve, please do a few fixes:
1. Remove Contributors from .gitignore
2. Remove dependencies of DotNetZip, NUnit.VisualStudio.TestAdapter from DotSpatial.Data.Tests and DotSpatial.Data

donogst commented:
Hi max,
I have updated my pr. I think it is ok now
On 18 Jan 2017 08:46, "Max Miroshnikov" <notifications@github.com> wrote:

> Steve, please do a few fixes:
>
>    1. Remove Contributors from .gitignore
>    2. Remove dependencies of DotNetZip, NUnit.VisualStudio.TestAdapter
>    from DotSpatial.Data.Tests and DotSpatial.Data
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DotSpatial/DotSpatial/pull/890#issuecomment-273416691>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFFvJ8lV9i7d-W6tkNbcM64Cc_Q4jpQEks5rTdF9gaJpZM4K3aeu>
> .
>

mogikanin commented:
@jany-tenaj please take a look into this PR too.

jany-tenaj commented:
@donogst Please correct the comments that were added. 

1. Use full sentences if possible and correct spelling. 
2. Explain clearly what the functions do, e.g. LineShapefile.SetHeaderShapeType comments that it returns something but it's a void function
3. If the function has parameters or return values please add comments for them, too.
4. Why are there two blanks after the . in the comment of Shapefile.WriteFileLength? I've seen this in other places, too, but still don't understand it.

In PointShapefile.ExportShapefilePackage decide whether you want to remove the `UpdateAttributes();` as indicated by your comment `// remove this call to update attributes //?` or remove the comment. Who is supposed to decide this if not you?

jany-tenaj commented:
@mogikanin While trying to fix the Export to work for !IndexMode I found the following thing.

We use [BufferedBinaryWriter ](https://github.com/DotSpatial/DotSpatial/blob/master/Source/DotSpatial.Data/LineShapefile.cs#L421) when not in IndexMode for Line, Polygon and Multipoint shapefiles.

We use [FileStream](https://github.com/DotSpatial/DotSpatial/blob/master/Source/DotSpatial.Data/PointShapefile.cs#L272) when not in IndexMode for point shapefiles and when in IndexMode for all shapefiles.

Do you have an idea why we're using BufferedBinaryWriter only in the first 3 cases? 

I'd like to change those 3 case to work like the others. Do you see any reason against doing that?

donogst commented:
Hi,

I think using bufferedbinarywriter is the part of the changes I made. The
idea was to not use filestream, to remove dependency on disks.


On 29 Mar 2017 10:40, "jany-tenaj" <notifications@github.com> wrote:

> @mogikanin <https://github.com/mogikanin> While trying to fix the Export
> to work for !IndexMode I found the following thing.
>
> We use BufferedBinaryWriter
> <https://github.com/DotSpatial/DotSpatial/blob/master/Source/DotSpatial.Data/LineShapefile.cs#L421>
> when not in IndexMode for Line, Polygon and Multipoint shapefiles.
>
> We use FileStream
> <https://github.com/DotSpatial/DotSpatial/blob/master/Source/DotSpatial.Data/PointShapefile.cs#L272>
> when not in IndexMode for point shapefiles and when in IndexMode for all
> shapefiles.
>
> Do you have an idea why we're using BufferedBinaryWriter only in the first
> 3 cases?
>
> I'd like to change those 3 case to work like the others. Do you see any
> reason against doing that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DotSpatial/DotSpatial/pull/890#issuecomment-290037995>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFFvJ5_qoWDzDtMVk15T4qOO_UnZdgDTks5rqib4gaJpZM4K3aeu>
> .
>

jany-tenaj commented:
@donogst Nope it's used in [master](https://github.com/DotSpatial/DotSpatial/blob/master/Source/DotSpatial.Data/LineShapefile.cs#L421), too.

mogikanin commented:
@jany-tenaj I think we should use standard `FileStream` in all cases, and do not use these custom BufferedBinaryWriter\BufferedBinaryReader. I don't see in them any sense, because `FileStream` itself uses buffering and we can specify bufferSize via 'FileStream' constructor,

jany-tenaj commented:
@mogikanin Shapefile class seems to be meant as base class for the specific shapefile types. Why does it have to be instantiateable? We already have FeatureSet that can be instantiated without specifying FeatureType or ShapeType of the features that will be added.

mogikanin commented:
Yes, we can make it abstract.

jany-tenaj commented:
@donogst Please check everything still works for you after my changes.

@mogikanin If you don't see any other problems I think we can merge now.

mogikanin commented:
@jany-tenaj I found small issue. After fixing it please merge PR.

mogikanin commented on Source/DotSpatial.Data/AttributeTable.cs:
```diff
@@ -888,6 +892,34 @@ public void Fill(int numRows)
+            try
+            {
+                using (var inMemoryStream = new MemoryStream())
+                using (_writer = new BinaryWriter(inMemoryStream))
```
Here should be used local variable, not field `_writer`.

jany-tenaj replied:
We have to use _writer because it is used inside WriteTable();

mogikanin replied:
Ah, ok. I think later we should refactor this class, because this `_writer` initialized in several places, and logic is messy,

jany-tenaj commented:
@mogikanin Are you going to fix it or tell me what the issue is?

mogikanin commented:
You should see it now. It is in "review".

jany-tenaj commented:
Sorry but I don't see any review.
